File: poker.cpp
#include <iostream>
#include <map>
#include <set>
#include <utility>
#include <vector>

using Players = std::vector<std::pair<std::string, std::string>>;
using Cards = std::set<std::string>;

std::map<int, int> kRating = {
    {'A', 14}, {'K', 13}, {'Q', 12}, {'J', 11}, {'T', 10}, {'9', 9}, {'8', 8},
    {'7', 7},  {'6', 6},  {'5', 5},  {'4', 4},  {'3', 3},  {'2', 2},
};

bool check_winner(const std::string& card, const Players& hands) {
  std::set<std::pair<int, int>> winners{};
  for (int i = 0; i < hands.size(); i++) {
    if (hands[i].first[0] == hands[i].second[0] &&
        hands[i].first[0] == card[0]) {
      winners.insert(std::make_pair(300 + kRating[hands[i].first[0]], i));
      continue;
    }
    if (hands[i].first[0] == hands[i].second[0] ||
        hands[i].first[0] == card[0]) {
      winners.insert(std::make_pair(200 + kRating[hands[i].first[0]], i));
      continue;
    }
    if (hands[i].second[0] == card[0]) {
      winners.insert(std::make_pair(200 + kRating[hands[i].second[0]], i));
      continue;
    }
    int max_card =
        std::max(kRating[hands[i].first[0]], kRating[hands[i].second[0]]);
    max_card = std::max(max_card, kRating[card[0]]);
    winners.insert(std::make_pair(max_card, i));
  }
  bool result = false;
  auto winner = *winners.rbegin();
  if (winner.second == 0) {
    result = true;
  } else {
    for (auto w = winners.rbegin();
         w != winners.rend() && w->first == winner.first; w++) {
      if (w->second == 0) {
        result = true;
        break;
      }
    }
  }
  return result;
}

int main(void) {
  Cards deck{"AS", "2S", "3S", "4S", "5S", "6S", "7S", "8S", "9S", "TS", "JS",
             "QS", "KS", "AH", "2H", "3H", "4H", "5H", "6H", "7H", "8H", "9H",
             "TH", "JH", "QH", "KH", "AD", "2D", "3D", "4D", "5D", "6D", "7D",
             "8D", "9D", "TD", "JD", "QD", "KD", "AC", "2C", "3C", "4C", "5C",
             "6C", "7C", "8C", "9C", "TC", "JC", "QC", "KC"};
  int t, n;
  std::cin >> t;
  std::vector<Players> hands(t);
  std::vector<Cards> deck_copy(t, deck);
  for (int j = 0; j < t; j++) {
    std::cin >> n;
    hands[j].resize(n);
    for (int i = 0; i < n; i++) {
      std::cin >> hands[j][i].first >> hands[j][i].second;
      deck_copy[j].erase(hands[j][i].first);
      deck_copy[j].erase(hands[j][i].second);
    }
  }

  for (int i = 0; i < t; i++) {
    std::vector<std::string> win_deck{};
    for (auto card : deck_copy[i]) {
      if (check_winner(card, hands[i])) {
        win_deck.push_back(card);
      }
    }
    std::cout << win_deck.size() << std::endl;
    for (auto card : win_deck) {
      std::cout << card << std::endl;
    }
  }
  return 0;
}

File: poker.py
kRating = {
    'A': 14, 'K': 13, 'Q': 12, 'J': 11, 'T': 10, '9': 9, '8': 8,
    '7': 7, '6': 6, '5': 5, '4': 4, '3': 3, '2': 2,
}


def check_winner(card, hands):
    winners = []
    for i in range(len(hands)):
        if hands[i][0][0] == hands[i][1][0] and hands[i][0][0] == card[0]:
            winners.append((300 + kRating[hands[i][0][0]], i))
            continue
        if hands[i][0][0] == hands[i][1][0] or hands[i][0][0] == card[0]:
            winners.append((200 + kRating[hands[i][0][0]], i))
            continue
        if hands[i][1][0] == card[0]:
            winners.append((200 + kRating[hands[i][1][0]], i))
            continue
        max_card = max(kRating[hands[i][0][0]],
                       kRating[hands[i][1][0]], kRating[card[0]])
        winners.append((max_card, i))
    result = False
    winners.sort(reverse=True)
    winner = winners[0]
    if winner[1] == 0:
        result = True
    else:
        for w in winners:
            if w[0] == winner[0] and w[1] == 0:
                result = True
                break
    return result


def main():
    deck = ["AS", "2S", "3S", "4S", "5S", "6S", "7S", "8S", "9S", "TS", "JS",
            "QS", "KS", "AH", "2H", "3H", "4H", "5H", "6H", "7H", "8H", "9H",
            "TH", "JH", "QH", "KH", "AD", "2D", "3D", "4D", "5D", "6D", "7D",
            "8D", "9D", "TD", "JD", "QD", "KD", "AC", "2C", "3C", "4C", "5C",
            "6C", "7C", "8C", "9C", "TC", "JC", "QC", "KC"]
    t = int(input())
    hands = []
    deck_copy = [deck.copy() for _ in range(t)]
    for test in range(t):
        n = int(input())
        player_hands = []
        for _ in range(n):
            hand = input().split()
            player_hands.append((hand[0], hand[1]))
            deck_copy[test].remove(hand[0])
            deck_copy[test].remove(hand[1])
        hands.append(player_hands)

    for i in range(t):
        win_deck = [card for card in deck_copy[i]
                    if check_winner(card, hands[i])]
        print(len(win_deck))
        for card in win_deck:
            print(card)


if __name__ == "__main__":
    main()
